subsys/mgmt/mcumgr/grp/settings_mgmt/src/settings_mgmt.c

```diff
@@ -0,0 +1,535 @@
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	/* Update the settings data */
+	rc = settings_runtime_set(key_name, data.value, data.len);
+
+	if (rc < 0) {
+		if (rc == -EINVAL) {
+			rc = SETTINGS_MGMT_RET_RC_KEY_NOT_FOUND;
+		} else {
+			rc = SETTINGS_MGMT_RET_RC_UNKNOWN;
+		}
+
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS, (uint16_t)rc);
+	}
+
+end:
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	free(key_name);
+#endif
+
+	return ok ? MGMT_ERR_EOK : MGMT_ERR_EMSGSIZE;
+}
+
+/**
+ * Command handler: settings delete
+ */
+static int settings_mgmt_delete(struct smp_streamer *ctxt)
+{
+	zcbor_state_t *zse = ctxt->writer->zs;
+	zcbor_state_t *zsd = ctxt->reader->zs;
+	int rc;
+	bool ok;
+	size_t decoded;
+	struct zcbor_string key = { 0 };
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	char *key_name = NULL;
+#else
+	char key_name[CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN];
+#endif
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	struct settings_mgmt_access settings_access_data = {
+		.access = SETTINGS_ACCESS_DELETE,
+		.name = key_name,
+	};
+
+	enum mgmt_cb_return status;
+	int32_t ret_rc;
+	uint16_t ret_group;
+#endif
+
+	struct zcbor_map_decode_key_val settings_delete_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("name", zcbor_tstr_decode, &key),
+	};
+
+	ok = zcbor_map_decode_bulk(zsd, settings_delete_decode, ARRAY_SIZE(settings_delete_decode),
+				   &decoded) == 0;
+
+	if (!ok || key.len == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Check if the length of the user supplied key is too large for the configuration
+	 * to allow and return an error if so
+	 */
+	if (key.len >= CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN) {
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS,
+				     SETTINGS_MGMT_RET_RC_KEY_TOO_LONG);
+		goto end;
+	}
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	key_name = (char *)malloc(key.len + 1);
+
+	if (key_name == NULL) {
+		return MGMT_ERR_ENOMEM;
+	}
+#endif
+
+	memcpy(key_name, key.value, key.len);
+	key_name[key.len] = 0;
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	/* Send request to application to check if access should be allowed or not */
+	status = mgmt_callback_notify(MGMT_EVT_OP_SETTINGS_MGMT_ACCESS, &settings_access_data,
+				      sizeof(settings_access_data), &ret_rc, &ret_group);
+
+	if (status != MGMT_CB_OK) {
+		bool ok;
+
+		if (status == MGMT_CB_ERROR_RC) {
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	/* Delete requested key from settings */
+	rc = settings_delete(key_name);
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	free(key_name);
+#endif
+
+	if (rc < 0) {
+		if (rc == -ENOENT) {
+			rc = SETTINGS_MGMT_RET_RC_KEY_NOT_FOUND;
+		} else {
+			rc = SETTINGS_MGMT_RET_RC_UNKNOWN;
+		}
+
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS, (uint16_t)rc);
+	}
+
+end:
+	return ok ? MGMT_ERR_EOK : MGMT_ERR_EMSGSIZE;
```

Comment:
We repeat this line so often that I start to wonder whether we should maybe have a macro for that.

Reply:
Added